### PR TITLE
[SPARK-30806][SQL]Evaluate once per group in UnboundedWindowFunctionFrame

### PR DESCRIPTION
**Pull request by WangGuangxin**
### What changes were proposed in this pull request?
We only need to do aggregate evaluation once per group in `UnboundedWindowFunctionFrame`

### Why are the changes needed?
Currently, in `UnboundedWindowFunctionFrame.write`，it re-evaluate the processor for each row in a group, which is not necessary in fact which I'll address later. It hurts performance when the evaluation is time-consuming (for example, Percentile's eval need to sort its buffer and do some calculation). In our production, there is a percentile with window operation sql,  it costs more than 10 hours in SparkSQL while 10min in Hive.

In fact, `UnboundedWindowFunctionFrame` can be treated as `SlidingWindowFunctionFrame` with `lbound = UnboundedPreceding` and `ubound = UnboundedFollowing`, just as its comments. In that case, `SlidingWindowFunctionFrame` also only do evaluation once for each group.

The performance issue can be reproduced by running the follow scripts in local spark-shell
```
spark.range(100*100).map(i => (i, "India")).toDF("uv", "country").createOrReplaceTempView("test")
sql("select uv, country, percentile(uv, 0.95) over (partition by country) as ptc95 from test").collect.foreach(println)
```
Before this patch, the sql costs **128048 ms**.
With this patch,  the sql costs **3485 ms**.

If we increase the data size to 1000*1000 for example, then spark cannot even produce result without this patch(I'v waited for several hours).

### Does this PR introduce any user-facing change?
NO

### How was this patch tested?
Existing UT

### CODE REVIEW

**Comment by WangGuangxin** [comment]
@cloud-fan @maropu  Could you please help review this?

**Comment by maropu** [comment]
Could you show us a simple query for the performance pitfall?

**Comment by WangGuangxin** [comment]
> Could you show us a simple query for the performance pitfall?

yes, this can be reproduced by

```
spark.range(1000*1000).map(i => (i, "India")).toDF("uv", "country").createOrReplaceTempView("test")
sql("select uv, country, percentile(uv, 0.95) over (partition by country) as ptc95 from test").collect.foreach(println)
```

**Comment by maropu** [comment]
Please update the PR description to put the query and the performance numbers before/after this PR?

**Comment by maropu** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala [hunk 1]
```diff
@@ -284,9 +284,7 @@ final class UnboundedWindowFunctionFrame(
 
   /** Write the frame columns for the current row to the given target row. */
   override def write(index: Int, current: InternalRow): Unit = {
-    // Unfortunately we cannot assume that evaluation is deterministic. So we need to re-evaluate
-    // for each row.
-    if (processor != null) {
+    if (processor != null && (index == 0)) {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala]: I can't remember why I added this. Are you sure that computing the result of an aggregate (on its buffer) is always deterministic? If so then I am ok with this change.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala]: so unbounded window frame means we only need to calculate result once and always return the result for each input row?

Then shall we call `processor.evaluate(target)` in `prepare` and do thing in `write`?

Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala]: That would also work :)

Comment (WangGuangxin) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala]: > I can't remember why I added this. Are you sure that computing the result of an aggregate (on its buffer) is always deterministic? If so then I am ok with this change.

I've checked the subclass of `AggregateFunction`, and didn't found non-deterministic agg. 

Logically speaking, only non-ordered window spec is resolved to unbounded window frame (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L2719), so if `processor.evaluate(target)` is various for each row in a same group, the result is always non-deterministic, right?

**Comment by WangGuangxin** [comment]
I checked the logic in hive, it evaluate only once for unbound frame. 
Related code is https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/udf/ptf/WindowingTableFunction.java#L110. 
If the frame is unbound, `processWindow` return false, so that it goes to `out = new SameList(iPart.size(), out);`

**Comment by cloud-fan** [comment]
can you address https://github.com/apache/spark/pull/27558#discussion_r378872179 ?

**Comment by WangGuangxin** [comment]
> can you address [#27558 (comment)](https://github.com/apache/spark/pull/27558#discussion_r378872179) ?

sure

**Review by cloud-fan — APPROVED** [review]

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by hvanhovell** [comment]
Merging to master/3.0.

**Pull request closed without merging**